base/src/org/compiere/model/MBPartner.java

```diff
@@ -861,6 +861,15 @@ public String getSOCreditStatus (BigDecimal additionalAmt)
 		creditLimit = creditLimit.subtract(additionalAmt);
 		if (creditLimit.compareTo(getTotalOpenBalance()) < 0)
 			return SOCREDITSTATUS_CreditHold;
+		

```

Comment:
I make this query with a explain for PostgreSQL and folow result:

```SQL
Aggregate  (cost=620.07..620.08 rows=1 width=8)
  ->  Hash Join  (cost=619.31..620.00 rows=30 width=0)
        Hash Cond: ("*SELECT* 1".c_paymentterm_id = pt.c_paymentterm_id)
        ->  HashAggregate  (cost=616.03..616.33 rows=30 width=1150)
              Group Key: "*SELECT* 1".ad_org_id, "*SELECT* 1".ad_client_id, "*SELECT* 1".documentno, "*SELECT* 1".c_invoice_id, "*SELECT* 1".c_order_id, "*SELECT* 1".c_bpartner_id, "*SELECT* 1".issotrx, "*SELECT* 1".dateinvoiced, "*SELECT* 1".dateacct, "*SELECT* 1".netdays, "*SELECT* 1".duedate, "*SELECT* 1".daysdue, (("*SELECT* 1".discountdate)::timestamp without time zone), "*SELECT* 1".discountamt, "*SELECT* 1".grandtotal, "*SELECT* 1".paidamt, "*SELECT* 1".openamt, "*SELECT* 1".c_currency_id, "*SELECT* 1".c_conversiontype_id, "*SELECT* 1".c_paymentterm_id, "*SELECT* 1".ispayschedulevalid, (NULL::numeric), "*SELECT* 1".invoicecollectiontype, "*SELECT* 1".c_campaign_id, "*SELECT* 1".c_project_id, "*SELECT* 1".c_activity_id, "*SELECT* 1".salesrep_id, "*SELECT* 1".description, "*SELECT* 1".c_doctype_id, "*SELECT* 1".c_bp_accounttype_id, "*SELECT* 1".c_bp_salesgroup_id, "*SELECT* 1".c_bp_segment_id, "*SELECT* 1".c_bp_industrytype_id, "*SELECT* 1".m_pricelist_id, "*SELECT* 1".c_salesregion_id
              ->  Append  (cost=88.65..613.40 rows=30 width=1150)
                    ->  Subquery Scan on "*SELECT* 1"  (cost=88.65..391.28 rows=29 width=427)
                          ->  Hash Join  (cost=88.65..390.92 rows=29 width=423)
                                Hash Cond: (i.c_paymentterm_id = p.c_paymentterm_id)
                                ->  Nested Loop  (cost=85.37..343.33 rows=29 width=850)
                                      ->  Hash Join  (cost=85.09..300.99 rows=29 width=271)
                                            Hash Cond: (i.c_bpartner_location_id = bpl.c_bpartner_location_id)
                                            ->  Nested Loop  (cost=20.62..236.44 rows=29 width=271)
                                                  ->  Index Scan using c_bpartner_pkey on c_bpartner b  (cost=0.28..8.29 rows=1 width=61)
                                                        Index Cond: (c_bpartner_id = '1000326'::numeric)
                                                  ->  Hash Join  (cost=20.34..227.86 rows=29 width=216)
                                                        Hash Cond: (i.c_doctype_id = d.c_doctype_id)
                                                        ->  Bitmap Heap Scan on c_invoice i  (cost=4.86..212.30 rows=29 width=212)
                                                              Recheck Cond: (c_bpartner_id = '1000326'::numeric)
                                                              Filter: ((ispayschedulevalid <> 'Y'::bpchar) AND (docstatus = ANY ('{CO,CL}'::bpchar[])) AND (issotrx = 'Y'::bpchar) AND (invoiceopen(c_invoice_id, '0'::numeric) <> '0'::numeric))
                                                              ->  Bitmap Index Scan on c_invoice_documentno_target  (cost=0.00..4.85 rows=76 width=0)
                                                                    Index Cond: (c_bpartner_id = '1000326'::numeric)
                                                        ->  Hash  (cost=11.88..11.88 rows=288 width=10)
                                                              ->  Seq Scan on c_doctype d  (cost=0.00..11.88 rows=288 width=10)
                                            ->  Hash  (cost=46.99..46.99 rows=1399 width=12)
                                                  ->  Seq Scan on c_bpartner_location bpl  (cost=0.00..46.99 rows=1399 width=12)
                                      ->  Index Only Scan using c_location_pkey on c_location loc  (cost=0.28..0.43 rows=1 width=6)
                                            Index Cond: (c_location_id = bpl.c_location_id)
                                ->  Hash  (cost=2.57..2.57 rows=57 width=15)
                                      ->  Seq Scan on c_paymentterm p  (cost=0.00..2.57 rows=57 width=15)
                    ->  Subquery Scan on "*SELECT* 2"  (cost=5.83..221.97 rows=1 width=440)
                          ->  Nested Loop  (cost=5.83..221.95 rows=1 width=412)
                                Join Filter: ((i_1.c_invoice_id = ips.c_invoice_id) AND (invoiceopen(i_1.c_invoice_id, ips.c_invoicepayschedule_id) <> '0'::numeric))
                                ->  Nested Loop  (cost=5.83..219.39 rows=1 width=850)
                                      ->  Nested Loop  (cost=5.56..217.93 rows=1 width=271)
                                            ->  Nested Loop  (cost=5.28..209.63 rows=1 width=271)
                                                  ->  Nested Loop  (cost=5.00..201.32 rows=1 width=216)
                                                        ->  Bitmap Heap Scan on c_invoice i_1  (cost=4.85..193.10 rows=1 width=212)
                                                              Recheck Cond: (c_bpartner_id = '1000326'::numeric)
                                                              Filter: ((docstatus = ANY ('{CO,CL}'::bpchar[])) AND (ispayschedulevalid = 'Y'::bpchar) AND (issotrx = 'Y'::bpchar))
                                                              ->  Bitmap Index Scan on c_invoice_documentno_target  (cost=0.00..4.85 rows=76 width=0)
                                                                    Index Cond: (c_bpartner_id = '1000326'::numeric)
                                                        ->  Index Scan using c_doctype_pkey on c_doctype d_1  (cost=0.15..8.17 rows=1 width=10)
                                                              Index Cond: (c_doctype_id = i_1.c_doctype_id)
                                                  ->  Index Scan using c_bpartner_pkey on c_bpartner b_1  (cost=0.28..8.29 rows=1 width=61)
                                                        Index Cond: (c_bpartner_id = '1000326'::numeric)
                                            ->  Index Scan using c_bpartner_location_pkey on c_bpartner_location bpl_1  (cost=0.28..8.29 rows=1 width=12)
                                                  Index Cond: (c_bpartner_location_id = i_1.c_bpartner_location_id)
                                      ->  Index Only Scan using c_location_pkey on c_location loc_1  (cost=0.28..0.43 rows=1 width=6)
                                            Index Cond: (c_location_id = bpl_1.c_location_id)
                                ->  Seq Scan on c_invoicepayschedule ips  (cost=0.00..1.02 rows=1 width=112)
                                      Filter: (isvalid = 'Y'::bpchar)
        ->  Hash  (cost=2.57..2.57 rows=57 width=6)
              ->  Seq Scan on c_paymentterm pt  (cost=0.00..2.57 rows=57 width=6)
```

**The cost is:** cost=620.07..620.08 rows=1 width=8
Some notes: 
- Join with payment term is unnecessary
- RV_OpenItem is a view very too lazy for load because use many functions for get current balance for invoice
- You only need if a invoice (unpaid) is due

I change query for other with same BP 

```SQL
SELECT 1 FROM C_Invoice i 
WHERE i.IsSotrx = 'Y' AND i.C_BPartner_ID=1000326
AND i.IsPaid = 'N'
AND (paymenttermDueDays(i.C_PaymentTerm_ID, i.DateInvoiced, getDate()) - **Here should be dunning grace**) > 0;

Note that it can be running with same result but more fast:

```SQL
 Hash Join  (cost=160.75..192.72 rows=1 width=4)
   Hash Cond: (i.c_paymentterm_id = pt.c_paymentterm_id)
   Join Filter: (((paymenttermduedays(i.c_paymentterm_id, (i.dateinvoiced)::timestamp with time zone, getdate()))::numeric - pt.gracedays) > '0'::numeric)
   ->  Bitmap Heap Scan on c_invoice i  (cost=157.47..189.42 rows=4 width=13)
         Recheck Cond: ((c_bpartner_id = '1000326'::numeric) AND (ispaid = 'N'::bpchar))
         Filter: (issotrx = 'Y'::bpchar)
         ->  BitmapAnd  (cost=157.47..157.47 rows=9 width=0)
               ->  Bitmap Index Scan on c_invoice_documentno_target  (cost=0.00..4.85 rows=76 width=0)
                     Index Cond: (c_bpartner_id = '1000326'::numeric)
               ->  Bitmap Index Scan on c_invoice_paid  (cost=0.00..152.37 rows=497 width=0)
                     Index Cond: (ispaid = 'N'::bpchar)
   ->  Hash  (cost=2.57..2.57 rows=57 width=9)
         ->  Seq Scan on c_paymentterm pt  (cost=0.00..2.57 rows=57 width=9)
```

**The cost is:** cost=160.75..192.72 rows=1 width=4



Here complete code for replace:

```Java
		//	Exist due invoices Above dunning Grace
		int noInvoicesDue = 0;
		StringBuffer sql = new StringBuffer("SELECT 1 "
				+ "FROM C_Invoice i "
				+ "LEFT JOIN C_PaymentTerm pt ON(pt.C_PaymentTerm_ID = i.C_PaymentTerm_ID) "
				+ "WHERE i.IsSotrx = 'Y' "
				+ "AND i.C_BPartner_ID=? "
				+ "AND i.IsPaid = 'N' "
				+ "AND (paymenttermDueDays(i.C_PaymentTerm_ID, i.DateInvoiced, getDate()) - pt.GraceDays) > 0");
		if(getDunningGrace() != null) {
			sql.append(" AND paymenttermDueDate(i.C_PaymentTerm_ID, i.DateInvoiced) >= ?");
			noInvoicesDue = DB.getSQLValueEx(null, sql.toString(), getC_BPartner_ID(), getDunningGrace());
		} else {
			noInvoicesDue = DB.getSQLValueEx(null, sql.toString(), getC_BPartner_ID());
		}
		if (noInvoicesDue == 1)
			return SOCREDITSTATUS_CreditHold;
```
Here result:

![Screenshot_20200222_212601](https://user-images.githubusercontent.com/2333092/75101989-cecc6080-55ba-11ea-8d9f-5982ad23c072.png)

Reply:
Thank you @yamelsenih .